program/src/processor.rs

```diff
@@ -644,3 +666,189 @@ pub fn process(program_id: &Pubkey, accounts: &[AccountInfo], input: &[u8]) -> P
+            // Evaluate.
+            if token_supply == 0 {
+                prop_assert_eq!(result, 0);
+            } else {
+                let expected = (rewards as u128)
+                    .checked_mul(REWARDS_PER_TOKEN_SCALING_FACTOR)
+                    .and_then(|product| product.checked_div(token_supply as u128))
+                    .unwrap();
+                prop_assert_eq!(result, expected);
+            }
+        }
+    }
+
+    // Since the marginal rate (current - last) within this function is describing
+    // rewards _per token_, which is calculated by `calculate_rewards_per_token`,
+    // the maximum permitted true reward difference on the system that can be
+    // tolerated by a maximum token account balance (100% of token supply) depends
+    // on the token supply of the mint.
+    //
+    // Consider the calculations for both functions:
+    //
+    //     rewards_per_token = (reward * 1e9) / mint_supply
+    //     eligible_rewards = (marginal_rewards_per_token * balance) / 1e9
+    //
+    // By rearranging the first formula (for `calculate_rewards_per_token`), we
+    // can calculate the total accumulated system-level reward based on the
+    // given rewards per token and mint supply.
+    //
+    //     accumulated_rewards =
+    //              (accumulated_rewards_per_token * mint_supply) / e9
+    //
+    // By rearranging the second formula (which is for this function
+    // `calculate_eligible_rewards`), and plugging in `u64::MAX * 1e9` for
+    // `eligible_rewards` (function upper bound), we can obtain a formula for
+    // the maximum marginal rate (current - last) supported.
+    //
+    //     max_marginal_rewards_per_token = (u64::MAX * 1e9) / balance
+    //
+    // The largest token account balance possible is 100% of the mint supply,
+    // so we can rewrite the rearranged second formula as follows.
+    //
+    //     max_marginal_rewards_per_token = (u64::MAX * 1e9) / mint_supply
+    //
+    // Now, plugging the formula for `max_marginal_rewards_per_token` into the
+    // `accumulated_rewards_per_token` of the rearranged first formula, we get
+    // the following.
+    //
+    //     max_accumulated_rewards = (
+    //                          ((u64::MAX * 1e9) / mint_supply) * mint_supply
+    //                        ) / 1e9
+    //
+    // As you can see, the mint supply can be factored out, and the max
+    // accumulated system-level reward can now be computed directly.
+    //
+    //     max_accumulated_rewards = u64::MAX
+    //
+    // Therefore, the maximum system-level accumulated reward cannot exceed
+    // `u64::MAX` without breaking `calculate_eligible_rewards`. Since
+    // accumulated rewards as stored as _per token_ values in a `u128`, it's
+    // mathematically possible to exceed `u64::MAX` in accumulated true
+    // rewards. However, note that `u64::MAX` exceeds the current circulating
+    // supply of SOL (`4.66e15`) by 386_266 %.
```

Comment:
Maybe I'm misunderstanding the terminology, but the total system-level accumulated reward can go over u64::MAX, especially if the rewards keep getting paid out over time. It's the marginal / undistributed amount that's capped to `u64::MAX`.

On a per-token basis though, if there's u64::MAX SOL and u64::MAX PAL, then the maximum marginal rewards per token is 10^9, aka 1 lamport per token. That's probably the most important bit

Reply:
I see what you're saying. I think I just went about this in a weird direction. Check out my updated version of the arithmetic. I think it makes more sense now.